### ISSUE TITLE
Call to intersection with union argument is incorrectly rejected

### ISSUE THREAD

**Issue by JelleZijlstra** [astral-sh/ty#3423]
### Summary

```python
from ty_extensions import Intersection
from typing import Callable, Any

def f(
    x1: Intersection[Callable[[int], str], Callable[[str], int]],
    int_str: int | str,
) -> None:
    reveal_type(x1(int_str))
```
Playground: https://play.ty.dev/0920605e-87c4-4fc1-a64b-e6dd88cb39da

Given a type `((int) -> str) & ((str) -> int)`, ty rejects a call with argument type `int | str`.

ty accepts the equivalent code with overloads.

**Comment by JelleZijlstra** [comment]
It's also strange that the revealed type here is `Never` rather than `Any`. Generally type errors should result in Any, and that's what ty does with other calls.

**Comment by carljm** [comment]
Thank you! Yes, we need to apply union expansion in intersection calls also, not just in overload calls.

> Generally type errors should result in Any, and that's what ty does with other calls.

I don't think this is generally true for failed calls, in ty or any other type checker. Type checkers generally make the assumption that a failed call still returns the annotated return type of the called function. E.g. every type checker infers `int` here, not `Any`:

```py
def f(x: int) -> int:
    return x

reveal_type(f("foo"))
```

But I agree that it is wrong to apply this precedent to failed intersection calls by intersecting the return types of all callable intersection elements, as it looks like we currently do. Probably in the intersection case `Any` is the right answer, since we have no idea which intersection element should be the "true" match. (This is another case where `AnyOf` would be an even better answer, if we had it available.)

**Comment by JelleZijlstra** [comment]
Another variant without an explicit union:

https://play.ty.dev/9aa32deb-5e0a-402a-8853-bddbab0032ef

```python
from typing import Any, reveal_type, Callable, Literal
from ty_extensions import Intersection

class A: pass
class B(A): pass
class C(A): pass


def f(cb: Intersection[
    Callable[[Literal[True]], B],
    Callable[[Literal[False]], C],
], arg: bool):
    reveal_type(cb(arg))
```

```
Revealed type: `B & C` (revealed-type) [Ln 13, Col 17]
Argument is incorrect: Expected `Literal[True]`, found `bool` (invalid-argument-type) [Ln 13, Col 20]
Argument is incorrect: Expected `Literal[False]`, found `bool` (invalid-argument-type) [Ln 13, Col 20]
```

And even more perniciously:

https://play.ty.dev/e7c44596-f7cf-4b57-b264-5994002bde1f

```python
from typing import Any, reveal_type, Callable, Literal
from ty_extensions import Intersection, Not

class A: pass
class B(A): pass
class C(A): pass


def f(cb: Intersection[
    Callable[[Intersection[int, Not[bool]]], B],  # (int & ~bool) -> B
    Callable[[bool], C],  # (bool) -> C
], arg: int):
    reveal_type(cb(arg))
```

```
Revealed type: `B & C` (revealed-type) [Ln 13, Col 17]
Argument is incorrect: Expected `int & ~bool`, found `int` (invalid-argument-type) [Ln 13, Col 20]
Argument is incorrect: Expected `bool`, found `int` (invalid-argument-type) [Ln 13, Col 20]
```

In both cases the inferred type should be `B | C`.

**Comment by JelleZijlstra** [comment]
Also, assigning to myself. I'm thinking about this in the context of figuring out how to specify intersection calls, but I don't think it's urgent to fix; complex intersection calls don't seem particularly relevant in practice. I'll try to work on this after I've worked out to my satisfaction how it should work.

**Comment by carljm** [comment]
Ultimately I think this is the same as the existing "union expansion" specified for overload calls (which includes expanding `bool` to `Literal[True]` and `Literal[False]`, thus covering your first example in https://github.com/astral-sh/ty/issues/3423#issuecomment-4519767898, and also includes e.g. expanding an enum type to the union of its elements). The obvious initial fix here would be to apply the same logic to intersection calls that we already do to overload calls.

There are certainly cases that could/should be expanded but aren't specified to be expanded; your second example is one (though it requires negation types to express). I think for this to be generally doable would require a different approach. Currently type checkers just "try all the expansions and see if any of them work"; for full generalization to be feasible it would probably require inspecting the overload arguments to determine what expansions have a chance of distinguishing overloads.

I suspect that whatever is specified for intersection calls should match what is specified for overload calls, in this area.

**Comment by JelleZijlstra** [comment]
I fully implemented this now in JelleZijlstra/pycroscope#510.

```
$ pycroscope -c '''from typing import Any, reveal_type, Callable, Literal
from pycroscope.extensions import Intersection, Not

class A: pass
class B(A): pass
class C(A): pass


def f(cb: Intersection[
    Callable[[Intersection[int, Not[bool]]], B],  # (int & ~bool) -> B
    Callable[[bool], C],  # (bool) -> C
], arg: int):
    reveal_type(cb(arg))
'''

Revealed type is '<code>.B | <code>.C' (code: reveal_type)
In <code> at line 13
  10:     Callable[[Intersection[int, Not[bool]]], B],  # (int & ~bool) -> B
  11:     Callable[[bool], C],  # (bool) -> C
  12: ], arg: int):
  13:     reveal_type(cb(arg))
                      ^
```

The general idea is that whenever we try to apply a signature, we have three options:
- The argument type is a subtype of the parameter type. We're done; the argument is accepted.
- The argument type is disjoint from the parameter type. We're also done; this signature can't be useful.
- There is partial overlap between the argument A and parameter P types. Then we split the argument type into `A & P` and `A & ~P`, and check both variants against the whole intersection.

(With further complications around gradual types.) We keep splitting until there's nothing left to split.

This same algorithm could indeed also be applied to overloads.
